### PR TITLE
Adding New-Service to -SecurityDescriptorSDDL parameter like Set-Service

### PR DESCRIPTION
**Pull request by kvprasoon**
# PR Summary

This PR is to add `-SecurityDescriptorSDDL` parameter to `New-Service` cmdlet so that its in par with `Set-Service` in setting sddl for a service.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/pull/4735
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs [hunk 1]
```diff
@@ -112,6 +112,40 @@ protected bool ShouldProcessServiceOperation(ServiceController service)
...
         }
 
+        internal void SetServiceSecurityDescriptor(
+            ServiceController service,
+            string securityDescriptorSddl,
+            NakedWin32Handle hService)
+        {
+            var rawSecurityDescriptor = new RawSecurityDescriptor(securityDescriptorSddl);
+            RawAcl rawDiscretionaryAcl  = rawSecurityDescriptor.DiscretionaryAcl ;
+            var  discretionaryAcl   = new DiscretionaryAcl (false, false, rawDiscretionaryAcl);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: Extra spaces in two lines above.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs [hunk 2]
```diff
@@ -112,6 +112,40 @@ protected bool ShouldProcessServiceOperation(ServiceController service)
...
+            byte[] rawDacl = new byte[discretionaryAcl.BinaryLength];
+            discretionaryAcl.GetBinaryForm(rawDacl, 0);
+            rawSecurityDescriptor.DiscretionaryAcl = new RawAcl(rawDacl, 0);
+            byte[] securityDescriptorByte = new byte[rawSecurityDescriptor.BinaryLength];
+            rawSecurityDescriptor.GetBinaryForm(securityDescriptorByte, 0);
+
+            bool status = NativeMethods.SetServiceObjectSecurity(
+                        hService,
+                        SecurityInfos.DiscretionaryAcl,
+                        securityDescriptorByte);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: Indentation must be 4 spaces.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs (hunk 1 above)
Comment (kvprasoon) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: Thanks, why didn't CodeFactor catch this ?

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs (hunk 2 above)
Comment (kvprasoon) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: Thanks, why didn't CodeFactor catch this ?

**Comment by adityapatwardhan** [comment]
@PoshChan please retry static

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs [hunk 3]
```diff
@@ -2273,8 +2294,14 @@ protected override void BeginProcessing()
                     }
                 }
 
+                if (!string.IsNullOrEmpty(SecurityDescriptorSddl))
```
Comment (adityapatwardhan) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: would this work?

```csharp
service = new ServiceController(Name);

if (!string.IsNullOrEmpty(SecurityDescriptorSddl))
{
    SetServiceSecurityDescriptor(service, SecurityDescriptorSddl, hService);
}

File: test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1 [hunk 4]
```diff
@@ -27,6 +27,28 @@ Describe "Set/New/Remove-Service cmdlet tests" -Tags "Feature", "RequireAdminOnW
             $testservice2 = New-Service -BinaryPathName $svcfullpath -Name $testservicename2 -DependsOn $testservicename1
             $testservice2 | Should -Not -BeNullOrEmpty
         }
+        Function CheckSecurityDescriptorSddl {
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1]: Please add a empty line before starting the new function
```suggestion

        function CheckSecurityDescriptorSddl {
```

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs (hunk 3 above)
Comment (vexx32) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: That's a parameter, right? Shouldn't we be using `if (MyInvocation.BoundParameters.ContainsKey(nameof(SecurityDescriptorSddl))) { ... }`

**Comment by adityapatwardhan** [comment]
@kvprasoon Please follow the template and file a doc issue for the new parameter. If it is filed, please paste the link here.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs (hunk 1 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: Sometimes it is not predictable :-)

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs (hunk 3 above)
Comment (kvprasoon) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs]: @adityapatwardhan @vexx32  yes, I've used the same way how its implemented for `Set-Service`. I referred

L1883 `if (!string.IsNullOrEmpty(Status))` which is used for `-Status` parameter in `Set-Service`.

**Comment by kvprasoon** [comment]
doc changes https://github.com/MicrosoftDocs/PowerShell-Docs/pull/4735

**Review by adityapatwardhan — APPROVED** [review]

**Comment by iSazonov** [comment]
@kvprasoon Thanks for your contribution! Sorry for delay.

**Pull request merged**